common/message_passing/client.h

```diff
@@ -0,0 +1,154 @@
+#ifndef CSCI6780_MESSAGE_PASSING_CLIENT_H
+#define CSCI6780_MESSAGE_PASSING_CLIENT_H
+
+#include <google/protobuf/message.h>
+#include <sys/socket.h>
+
+#include <atomic>
+#include <condition_variable>
+#include <cstdint>
+#include <memory>
+#include <mutex>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "queue/queue.h"
+#include "tasks/receiver_task.h"
+#include "tasks/sender_task.h"
+#include "thread_pool/thread_pool.h"
+#include "types.h"
+#include "wire_protocol/wire_protocol.h"
+
+namespace message_passing {
+
+/**
+ * @brief Handles sending messages to another node.
+ */
+class Client {
+ public:
+  /**
+   * @param thread_pool Thread pool to use internally for managing associated
+   *    tasks.
+   * @param destination The destination node that we will send messages to.
+   */
+  Client(std::shared_ptr<thread_pool::ThreadPool> thread_pool,
+                Endpoint destination);
+  ~Client();
+
+  /**
+   * @brief Sends a message to another node synchronously. Will block until the
+   *    message is sent.
+   * @param message The message to send.
+   * @return Result of the internal call to `send()`.
+   */
+  int Send(const google::protobuf::Message& message);
+
+  /**
+   * @brief Sends a message to another node asynchronously. Will return
+   *    immediately, before the message is sent. Errors will not be reported.
+   * @param message The message to send.
+   * @return True if it succeeded in dispatching the send request, false
+   *    otherwise.
+   */
+  bool SendAsync(const google::protobuf::Message& message);
+
+  /**
+   * @brief Receives a message.
+   * @tparam MessageType The type of message we expect.
+   * @param message[out] The message that we received.
+   * @return True if a message was successfully received and parsed, false
+   *    otherwise.
+   */
+  template <class MessageType>
+  bool Receive(MessageType* message) {
```

Comment:
Why does the implementation live in the header file? I'm guessing it's because of the templating...

Reply:
Yep. Templated classes and functions, unfortunately, must be defined within the header file.